docstring/csc108_docstring.py

```diff
@@ -1,288 +1,94 @@
-        self.consume_token_type_or_error('}')
 
-    def parse_tuple(self):
-        """Parses a tuple recursively. Raises an exception if it is not a
-        proper docstring.
-        """
-        self.consume_token_type_or_error('(')
-        node = DocNode('tuple')
-        self.add_as_child_node(node)
-        self.push_then_parse_then_pop(node)
-        while self.get_current_token() == ',':
-            self.consume_token_type_or_error(',')
-            self.push_then_parse_then_pop(node)
-        self.consume_token_type_or_error(')')
+def to_simple_type(token):
+    s = token.string
+    if s == 'int':
+        return int
+    elif s == 'str':
+        return str
+    elif s == 'float':
+        return float
+    elif s == 'bool':
+        return bool
+    elif s in ['obj', 'object']:
+        return typing.Any
+    elif s in ['None', 'NoneType']:
+        return type(None)
+    return s
 
-    def consume_token_type_or_error(self, token_str):
-        """Consumes the token that is provided in the argument. If the token
-        is not found, then it will throw a NotA108DocstringException. The
-        stream will be advanced.
-        token_str: The string that should be found on the current token in the
-        stream.
-        """
-        if self.token_index >= len(self.tokens):
-            raise NotA108DocstringException('Expected "%s", terminated early' % token_str)
-        token = self.get_current_token()
-        if token != token_str:
-            raise NotA108DocstringException('Expected "%s", got "%s" instead' % (token_str, token))
-        self.advance_to_next_token()
 
-    def push_then_parse_then_pop(self, node):
-        """A convenience method that will push the node onto the stack, then
-        parse the element recursively, then pop this node off. This is used to
-        push a node that has child definitions onto the stack so it is
-        assigned the children appropriately, and then cleans up the stack.
-        node: The DocNode to go on the stack.
-        """
-        self.push_node_on_stack(node)
-        self.parse_element()
-        self.pop_node_stack()
+# -----------------------------------------------------------------------------
 
-    def get_current_token(self):
-        """Gets the current token."""
-        return self.tokens[self.token_index]
+element = forward_decl()
+elements = forward_decl()
 
-    def advance_to_next_token(self):
-        """Advances to the next token. Note: This does not throw an exception
-        if there is not a following token, the pointer is incremented."""
-        self.token_index += 1
+skip_of = skip_str('of')
+any_class = some(lambda x: True) >> to_simple_type
 
-    def add_as_child_node(self, node):
-        """Adds a node as a child to the last element on the node stack.
-        node: The node to add.
-        """
-        self.node_stack[-1].children.append(node)
+set_parser = (a_str('set') + skip_of + element) >> compile_set_type
```

Comment:
So if you think about it, I think for these parsers (including ones below), the type keywords like `set` and `list` can be `skip_str`'d as well. This would make the type of the "compile" functions a bit nicer.